tools/apiview/parsers/cpp-api-parser/ApiViewProcessor/AstNode.cpp

```diff
@@ -1318,6 +1318,45 @@ static std::string GetNavigationId(clang::FunctionDecl const* func)
   return navigationId;
 }
 
+static std::string GetNavigationId(
+    clang::FriendDecl const* friendDecl,
+    std::string const& parentNodeNavigationId)
+{
+  std::string navigationId = parentNodeNavigationId + "_friend_";
+  auto dc = friendDecl->getFriendDecl();
+  if (dc)
+  {
+    if (isa<FunctionDecl>(dc))
+    {
+      auto fd = cast<FunctionDecl>(dc);
+      navigationId += GetNavigationId(fd);
+    }
+    else if (isa<CXXRecordDecl>(dc))
+    {
+      auto rd = cast<CXXRecordDecl>(dc);
+      navigationId += rd->getQualifiedNameAsString();
+    }
+    else
+    {
+      dc->dump(llvm::outs());
+    }
+  }
+  else
+  {
+    auto friendType = friendDecl->getFriendType();
+    if (friendType)
+    {
+      navigationId
+          += QualType::getAsString(friendDecl->getFriendType()->getType().split(), LangOptions{});
+    }
+    else
+    {
+      friendDecl->dump(llvm::outs());
+    }
+  }
+  return navigationId;
```

Comment:
If you end up in the final else of the initial if/else then you could end up with navigationId being `<parentNodeId>_friend_`. Is a navigationId ending in "_" a valid format?

Reply:
All strings are valid, the only restriction is that the names must be unique.